pkg/policy/mapstate.go

```diff
@@ -74,24 +79,100 @@ type MapState interface {
 
 // mapState is a state of a policy map.
 type mapState struct {
-	allows mapStateMap
-	denies mapStateMap
+	allows *mapStateMap
+	denies *mapStateMap
+}
+
+type Identities interface {
+	GetNetsLocked(identity.NumericIdentity) []*net.IPNet
 }
 
 // mapStateMap is a convience type representing the actual structure mapping
 // policymap keys to policymap entries.
-type mapStateMap map[Key]MapStateEntry
+type mapStateMap struct {
+	t   bitlpm.Trie[bitlpm.Key[Key], map[uint32]MapStateEntry]
+	len int
+}
 
-func (m mapStateMap) insert(k Key, e MapStateEntry) {
-	if m == nil {
-		n := make(mapStateMap)
-		m = n
+// hardLookup returns the idMap associated with a key, but only returns
+// if the key prefixes match exactly.
+func (msm *mapStateMap) hardLookup(k Key) (idMap map[uint32]MapStateEntry, ok bool) {
```

Comment:
nit: Could be another candidate for integrating into the underlying datatype. Seems like this is a `LookupExact` or `LookupRange` with a tight bound on the range.

Reply:
The bitlpm library does some funky things with its interfaces. It's actually not possible to know the Key's specific prefix on lookup (which is why it is a parameter in the `Ancestors` and `Descendants` methods). The `Lookup` method in bitlpm works by looking up the Trie's maximum prefix.

Reply:
https://github.com/cilium/cilium/pull/32609